ocaml/xapi/repository.ml

```diff
@@ -186,50 +199,142 @@ let sync ~__context ~self ~token ~token_id =
-        let Pkg_mgr.{cmd; params} = Pkgs.make_cache ~repo_name in
-        ignore (Helpers.call_script cmd params) ;
 
         (* Sync with remote repository *)
-        let Pkg_mgr.{cmd; params} = Pkgs.sync_repo ~repo_name in
-        Unixext.mkdir_rec !Xapi_globs.local_pool_repo_dir 0o700 ;
-        clean_yum_cache repo_name ;
-        ignore (Helpers.call_script cmd params)
+        let sync_repo () =
+          let Pkg_mgr.{cmd; params} = Pkgs.sync_repo ~repo_name in
+          Unixext.mkdir_rec !Xapi_globs.local_pool_repo_dir 0o700 ;
+          clean_yum_cache repo_name ;
+          ignore (Helpers.call_script cmd params)
+        in
+        match Db.Repository.get_origin ~__context ~self with
+        | `remote ->
+            with_access_token ~token ~token_id @@ fun token_path ->
+            (* Configure proxy and token *)
+            let token_param =
+              match token_path with
+              | Some p ->
+                  Printf.sprintf "--setopt=%s.accesstoken=file://%s" repo_name p
+              | None ->
+                  ""
+            in
+            let proxy_url_param, proxy_username_param, proxy_password_param =
+              get_proxy_params ~__context repo_name
+            in
+            let Pkg_mgr.{cmd; params} =
+              [
+                "--save"
+              ; proxy_url_param
+              ; proxy_username_param
+              ; proxy_password_param
+              ; token_param
+              ]
+              |> fun config -> Pkgs.config_repo ~repo_name ~config
+            in
+            ignore
+              (Helpers.call_script ~log_output:Helpers.On_failure cmd params) ;
+            make_cache () ;
+            sync_repo ()
+        | `bundle ->
+            make_cache () ; sync_repo ()
+        | `remote_pool ->
+            let verified_rpc =
+              ext_host_verified_rpc ~__context ~cert remote_addr
+            in
+            let session_id =
+              try
+                Client.Client.Session.login_with_password ~rpc:verified_rpc
+                  ~uname:username ~pwd:password
+                  ~version:Datamodel_common.api_version_string
+                  ~originator:Xapi_version.xapi_user_agent
+              with
+              | Http_client.Http_request_rejected _ | Http_client.Http_error _
+              ->
+                raise
+                  (Api_errors.Server_error
+                     ( Api_errors
+                       .update_syncing_remote_pool_coordinator_service_failed
+                     , []
+                     )
+                  )
+            in
+            let xapi_token = session_id |> Ref.string_of in
+            with_xapi_token ~xapi_token @@ fun token_path ->
+            (* Configure xapi token *)
+            let token_param =
+              match token_path with
+              | Some p ->
+                  Printf.sprintf "--setopt=%s.xapitoken=file://%s" repo_name p
```

Comment:
I can imagine that `p` needs some kind of escaping when it contains spaces and other special characters. This is probably not an issue in practice; it might be still better to use the URI module to create this.